src/sdam/topology.ts

```diff
@@ -1112,4 +983,73 @@ function isRetryableWritesSupported(topology: any) {
   return true;
 }
 
+class ServerCapabilities {
+  constructor(ismaster: any) {
+    // Capabilities
+    let aggregationCursor = false;
+    let writeCommands = false;
+    let textSearch = false;
+    let authCommands = false;
+    let listCollections = false;
+    let listIndexes = false;
+    let maxNumberOfDocsInBatch = ismaster.maxWriteBatchSize || 1000;
+    let commandsTakeWriteConcern = false;
+    let commandsTakeCollation = false;
+
+    if (ismaster.minWireVersion >= 0) {
+      textSearch = true;
+    }
+
+    if (ismaster.maxWireVersion >= 1) {
+      aggregationCursor = true;
+      authCommands = true;
+    }
+
+    if (ismaster.maxWireVersion >= 2) {
+      writeCommands = true;
+    }
+
+    if (ismaster.maxWireVersion >= 3) {
+      listCollections = true;
+      listIndexes = true;
+    }
+
+    if (ismaster.maxWireVersion >= 5) {
+      commandsTakeWriteConcern = true;
+      commandsTakeCollation = true;
+    }
+
+    // If no min or max wire version set to 0
+    if (ismaster.minWireVersion == null) {
+      ismaster.minWireVersion = 0;
+    }
+
+    if (ismaster.maxWireVersion == null) {
+      ismaster.maxWireVersion = 0;
+    }
+
+    function setup_get_property(object: any, name: any, value: any) {
+      Object.defineProperty(object, name, {
+        enumerable: true,
+        get() {
+          return value;
+        }
+      });
+    }
+
+    // Map up read only parameters
+    setup_get_property(this, 'hasAggregationCursor', aggregationCursor);
+    setup_get_property(this, 'hasWriteCommands', writeCommands);
+    setup_get_property(this, 'hasTextSearch', textSearch);
+    setup_get_property(this, 'hasAuthCommands', authCommands);
+    setup_get_property(this, 'hasListCollectionsCommand', listCollections);
+    setup_get_property(this, 'hasListIndexesCommand', listIndexes);
+    setup_get_property(this, 'minWireVersion', ismaster.minWireVersion);
+    setup_get_property(this, 'maxWireVersion', ismaster.maxWireVersion);
+    setup_get_property(this, 'maxNumberOfDocsInBatch', maxNumberOfDocsInBatch);
+    setup_get_property(this, 'commandsTakeWriteConcern', commandsTakeWriteConcern);
+    setup_get_property(this, 'commandsTakeCollation', commandsTakeCollation);
```

Comment:
I think we should change these to declarative `get xxx() {}` functions which will improve the typing information

Reply:
Two thoughts:
- this class must be removed soon, it's a bad class that is being used to make decisions about command construction based on a random server description of a server in the topology.
- converting these to getters will require storing the `ismaster` passed in on the type

This isn't new code, even though it looks like it is. I think we should leave this alone, and work on a better replacement for it in future work.